pr_agent/algo/pr_processing.py

```diff
@@ -208,20 +208,47 @@ def pr_generate_compressed_diff(top_langs: list, token_handler: TokenHandler, mo
 
 
 async def retry_with_fallback_models(f: Callable):
-    model = get_settings().config.model
-    fallback_models = get_settings().config.fallback_models
-    if not isinstance(fallback_models, list):
-        fallback_models = [fallback_models]
-    all_models = [model] + fallback_models
-    for i, model in enumerate(all_models):
+    all_models = _get_all_models()
+    all_deployments = _get_all_deployments(all_models)
+    # try each (model, deployment_id) pair until one is successful, otherwise raise exception
+    for i, (model, deployment_id) in enumerate(zip(all_models, all_deployments)):
         try:
+            get_settings().set("openai.deployment_id", deployment_id)
             return await f(model)
         except Exception as e:
-            logging.warning(f"Failed to generate prediction with {model}: {traceback.format_exc()}")
+            logging.warning(
+                f"Failed to generate prediction with {model}"
+                f"{(' from deployment ' + deployment_id) if deployment_id else ''}: "
+                f"{traceback.format_exc()}"
+            )
             if i == len(all_models) - 1:  # If it's the last iteration
                 raise  # Re-raise the last exception
 
 
+def _get_all_models() -> List[str]:
+    model = get_settings().config.model
+    fallback_models = get_settings().config.fallback_models
+    if not isinstance(fallback_models, list):
+        fallback_models = [m.strip() for m in fallback_models.split(",")]
+    all_models = [model] + fallback_models
```

Comment:
**Suggestion:** Consider adding error handling for the case where the 'fallback_models' setting is not a string or a list. This will prevent unexpected errors if the setting is of a different type.
```suggestion
    fallback_models = get_settings().config.fallback_models
    if isinstance(fallback_models, str):
        fallback_models = [m.strip() for m in fallback_models.split(",")]
    elif not isinstance(fallback_models, list):
        raise ValueError('fallback_models setting must be a string or a list')
```